pkg/alertmanager/distributor/distributor.go

```diff
@@ -0,0 +1,306 @@
+		d.amSends.WithLabelValues(am.Addr).Inc()
+
+		// Use a background context to make sure all alertmanagers get the request even if we return early.
+		localCtx, cancel := context.WithTimeout(context.Background(), d.cfg.RemoteTimeout)
+		defer cancel()
+		localCtx = user.InjectOrgID(localCtx, userID)
+		if sp := opentracing.SpanFromContext(r.Context()); sp != nil {
+			localCtx = opentracing.ContextWithSpan(localCtx, sp)
+		}
+		// Get clientIP(s) from Context and add it to localCtx
+		localCtx = util.AddSourceIPsToOutgoingContext(localCtx, source)
+
+		resp, err := d.doRequest(localCtx, am, r, body)
+		if err != nil {
+			d.amSendFailures.WithLabelValues(am.Addr).Inc()
+			return err
+		}
+
+		if resp.GetStatus() != alertmanagerpb.OK {
+			return errors.New("alertmanager grpc request not ok")
+		}
+
+		return nil
+	}, func() {})
+
+	if err == nil {
+		w.WriteHeader(http.StatusOK)
+		return
+	}
+
+	d.respondFromError(err, w)
+}
+
+func (d *Distributor) doRead(userID string, w http.ResponseWriter, r *http.Request) {
+	key := shardByUser(userID)
+	replicationSet, err := d.alertmanagerRing.Get(key, RingOp, nil, nil, nil)
+	if err != nil {
+		level.Error(d.logger).Log("msg", "failed to get replication set from the ring", "err", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Until we have a mechanism to combine the results from multiple alertmanagers,
+	// we forward the request to only only of the alertmanagers.
+	amDesc := replicationSet.Ingesters[0]
+
+	var body []byte
+	if r.Body != nil {
+		body, err = ioutil.ReadAll(r.Body)
+		if err != nil {
+			level.Error(d.logger).Log("msg", "failed to read the request body during reads", "err", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	}
+
+	d.amSends.WithLabelValues(amDesc.Addr).Inc()
+	resp, err := d.doRequest(r.Context(), amDesc, r, body)
+	if err != nil {
+		d.amSendFailures.WithLabelValues(amDesc.Addr).Inc()
+		d.respondFromError(err, w)
+		return
+	}
+
+	if resp.GetStatus() != alertmanagerpb.OK {
+		http.Error(w, resp.Error, http.StatusInternalServerError)
+		return
+	}
+
+	http.Error(w, string(resp.HttpResponse.Body), int(resp.HttpResponse.Code))
+}
+
+func (d *Distributor) respondFromError(err error, w http.ResponseWriter) {
+	httpResp, ok := httpgrpc.HTTPResponseFromError(errors.Cause(err))
+	if !ok {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	for _, h := range httpResp.Headers {
+		for _, v := range h.Values {
+			w.Header().Add(h.Key, v)
+		}
+	}
+	http.Error(w, string(httpResp.Body), int(httpResp.Code))
+}
+
+func (d *Distributor) doRequest(ctx context.Context, am ring.InstanceDesc, r *http.Request, requestBody []byte) (*alertmanagerpb.Response, error) {
+	amClient, err := d.alertmanagerClientsPool.GetClientFor(am.Addr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get alertmanager from pool %s", am.Addr)
+	}
+
+	resp, err := amClient.HandleRequest(ctx, &alertmanagerpb.Request{
```

Comment:
If you use `server.HTTPRequest` from [`weaveworks/common`](https://github.com/weaveworks/common/blob/master/httpgrpc/server/server.go#L157-L168) you don't have to a) read the body initially and instead just copy the req, b) have `httpTogrpchttpHeaders` and c) manually construct the `httpgrpc.HTTPRequest` struct

Reply:
Yeah, I can do that for reads for now. Writes will require some additional overheads of creating another body reader and eventually consuming it again (more allocations), so for writes I think copying the body and not using `server.HTTPRequest` is better.